doc/developer_guide/developers.rst

```diff
@@ -495,6 +500,219 @@ DataAccess class i.e. the `_field_write_arguments()` and
+specific method implemented will call this. To see the
+class hierarchy, the following code can be written:
+
+::
+
+   from __future__ import print_function
+    class PrintHierarchy(PSyIRVisitor):
+        ''' Example of a visitor that prints the PSyIR node hierarchy. '''
+
+        def node(self, node):
+	    ''' This method is called if no specific methods have been
+	        written. '''
+	    print("[ {0} start]".format(type(node).__name__))
+	    for child in node.children:
+	        self.visit(child)
+	    print("[ {0} end]".format(type(node).__name__))
+
+    print_hierarchy = PrintHierarchy()
+    print_hierarchy.visit(psyir_tree)
+
+In the examples presented up to now, the information from a back-end
+has been printed. However, a back-end will generally not want to use
+print statements. Output from a `PSyIRVisitor` is supported by
+allowing each method call to return a string. Reimplementing the
+previous example using strings would give the following:
+
+::
+   
+    from __future__ import print_function class
+    PrintHierarchy(PSyIRVisitor):
+        ''' Example of a visitor that prints the PSyIR node hierarchy'''
+
+        def node(self, node):
+            ''' This method is called if the visitor finds a loop '''
+	    result = "[ {0} start ]".format(type(node).__name__)
+	    for child in node.children:
+	        result += self.visit(child)
+	    result += "[ {0} end ]".format(type(node).__name__)
+            return result
+
+    print_hierarchy = PrintHierarchy()
+    result = print_hierarchy.visit(psyir_tree)
+    print(result)
+
+As most back-ends are expected to indent their output based in some
+way on the PSyIR node hierarchy, the `PSyIRVisitor` provides support
+for this. The `self._nindent` variable contains the current
+indentation as a string and the indentation can be increased by
+increasing the value of the `self._depth` variable. The initial depth
+defaults to 0 and the initial indentation defaults to two
+spaces. These defaults can be changed when creating the back-end
+instance. For example:
+
+::
+
+    print_hierarchy = PrintHierarchy(initial_indent_depth=2,
+                                     indent_string="***")
+
+The `PrintHierarchy` example can be modified to support indenting by
+writing the following:
+
+::
+
+    from __future__ import print_function
+    class PrintHierarchy(PSyIRVisitor):
+    ''' Example of a visitor that prints the PSyIR node hierarchy
+        with indentation'''
+
+        def node(self, node):
+            ''' This method is called if the visitor finds a loop '''
+	    result = "{0}[ {1} start ]\n".format(self._nindent,
+	                                         type(node).__name__)
+	    self._depth += 1
+	    for child in node.children:
+	        result += self.visit(child)
+	    self._depth -= 1
+	    result += "{0}[ {1} end ]\n".format(self._nindent,
+	                                        type(node).__name__)
+            return result
```

Comment:
This statement has a bad indentation

Reply:
That's weird. It looks OK to me. I see that earlier lines start with a tab and this one does not so I've made this one start with a tab as well. Perhaps that will fix the problem.